pkg/controller/common/rollout/rollout_webhook.go

```diff
@@ -4,94 +4,107 @@ import (
 	"bytes"
 	"context"
 	"encoding/json"
-	"errors"
 	"fmt"
 	"io/ioutil"
 	"net/http"
 	"net/url"
-	"strconv"
-	"time"
+
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
 
 	"github.com/oam-dev/kubevela/apis/standard.oam.dev/v1alpha1"
+	"github.com/oam-dev/kubevela/pkg/controller/common"
 )
 
-func callWebhook(webhook string, payload interface{}, timeout string) error {
+// issue an http call to the an end ponit
+func makeHTTPRequest(ctx context.Context, webhookEndPoint string, payload interface{}) ([]byte, int, error) {
 	payloadBin, err := json.Marshal(payload)
 	if err != nil {
-		return err
+		return nil, http.StatusInternalServerError, err
 	}
 
-	hook, err := url.Parse(webhook)
+	hook, err := url.Parse(webhookEndPoint)
 	if err != nil {
-		return err
+		return nil, http.StatusInternalServerError, err
 	}
 
 	req, err := http.NewRequestWithContext(context.Background(), "POST", hook.String(), bytes.NewBuffer(payloadBin))
 	if err != nil {
-		return err
+		return nil, http.StatusInternalServerError, err
 	}
-
 	req.Header.Set("Content-Type", "application/json")
 
-	if timeout == "" {
-		timeout = "10s"
-	}
-
-	t, err := time.ParseDuration(timeout)
-	if err != nil {
-		return err
-	}
-
-	ctx, cancel := context.WithTimeout(req.Context(), t)
-	defer cancel()
-
-	r, err := http.DefaultClient.Do(req.WithContext(ctx))
+	// issue request with retry
+	var r *http.Response
+	var body []byte
+	err = retry.OnError(retry.DefaultBackoff,
+		func(error) bool {
+			// not sure what not to retry on
```

Comment:
maybe the error code 500 that we should not retry on.

Reply:
I think we need to retry 5xx but not 4xx.  That's not determined here though.